fiftyone/utils/hrm2.py

```diff
@@ -0,0 +1,1947 @@
+
+
+def apply_hrm2_to_dataset_as_groups(
+    model: "HRM2Model",
+    dataset: "fo.Dataset",
+    label_field: str = "human_pose",
+    batch_size: int = 1,
+    num_workers: int = 4,
+    image_slice_name: str = "image",
+    scene_slice_name: str = "3d",
+    output_dir: Optional[str] = None,
+) -> "fo.Dataset":
+    """Apply HRM2 model to a dataset and create grouped samples.
+
+    This function converts an image dataset into a grouped dataset where each
+    group contains:
+    - An image slice with the original image and 2D keypoints
+    - A 3D slice with the reconstructed mesh scene and 3D data
+
+    This is the standard FiftyOne pattern for linking 2D images with 3D
+    reconstructions and enables 3D visualization in the FiftyOne App.
+
+    Args:
+        model: an HRM2Model instance
+        dataset: the FiftyOne dataset to process
+        label_field (str): base name for label fields (will create
+            "{label_field}_2d" on image samples and "{label_field}_3d" on
+            scene samples)
+        batch_size (int): batch size for inference
+        num_workers (int): number of workers for data loading
+        image_slice_name (str): name for the image slice in groups
+        scene_slice_name (str): name for the 3D scene slice in groups
+        output_dir (None): directory for .fo3d scenes and meshes. If None,
+            defaults to fo.config.model_zoo_dir/hrm2. Pass False to disable
+            file export (in-memory labels only)
+
+    Returns:
+        the grouped dataset
+    """
+    # Set default output_dir if not specified
+    if output_dir is None:
+        output_dir = os.path.join(fo.config.model_zoo_dir, "hrm2")
+        import fiftyone.core.storage as fos
+
+        output_dir = fos.normalize_path(output_dir)
+    elif output_dir is False:
+        output_dir = None  # Disable export
+
+    # Check if dataset is already grouped
+    if dataset.media_type == "group":
+        logger.warning(
+            "Dataset is already grouped. This will add new slices to "
+            "existing groups."
+        )
+        is_already_grouped = True
+    else:
+        is_already_grouped = False
+
+    # Store original sample info before clearing
+    logger.info(f"Processing {len(dataset)} samples...")
+    original_data = []
+    for sample in dataset.iter_samples():
+        original_data.append(
+            {
+                "filepath": sample.filepath,
+                "id": sample.id,
+                "fields": {
+                    k: sample[k]
+                    for k in sample.field_names
+                    if k not in ["id", "filepath", "metadata", "_media_type"]
+                },
+            }
+        )
+
+    # Apply model to get predictions
+    dataset.apply_model(
+        model,
+        label_field=label_field,
+        batch_size=batch_size,
+        num_workers=num_workers,
+        output_dir=output_dir,
+    )
+
+    # Collect predictions with scene paths
+    predictions = []
```

Comment:
this can be expressed idiomatically as:

```py
predictions = dataset.values(label_field)
```

^much more efficient than iterating over the entire dataset(!)